docs/USAGE-GUIDE.md

```diff
@@ -500,6 +500,224 @@ legacy SHA-1 algorithms in CertificateVerify messages if TLS1.2 has been negotia
+how to handle C signals, or simply ignore the SIGPIPE signal by calling
+`signal(SIGPIPE, SIG_IGN)` before calling any s2n-tls IO methods.
+
+### Blocking or Non-Blocking?
+
+s2n-tls supports both blocking and non-blocking I/O.
+* In blocking mode, each s2n-tls I/O function will not return until it has sent
+or received all data.
+* In non-blocking mode an s2n-tls I/O function may return while there is
+still I/O pending. In this case `s2n_error_get_type()` will report **S2N_ERR_T_BLOCKED**
+and the value of the s2n_blocked_status parameter will be set appropriately. s2n-tls I/O functions
+should be called repeatedly until the **blocked** parameter is **S2N_NOT_BLOCKED**.
+
+Servers in particular usually prefer non-blocking mode. In blocking mode, a single connection
+blocks the thread while waiting for more IO. In non-blocking mode, multiple connections
+can make progress by returning control while waiting for more IO using methods like
+[`poll`](https://linux.die.net/man/2/poll) or [`select`](https://linux.die.net/man/2/select).
+
+To use s2n-tls in non-blocking mode, set the underlying file descriptors as non-blocking. 
+For example:
+```c
+int flags = fcntl(fd, F_GETFL, 0);
+if (flags < 0) return S2N_FAILURE;
+if (fcntl(fd, F_SETFL, flags | O_NONBLOCK) < 0) return S2N_FAILURE;
+```
+
+### Errors and Alerts
+
+If the peer sends an alert, the next call to a read IO method will report **S2N_FAILURE** and
+`s2n_error_get_type()` will return **S2N_ERR_T_ALERT**. The specific alert received
+is available by calling `s2n_connection_get_alert()`.
+
+In TLS1.3, all alerts are fatal. s2n-tls also treats all alerts as fatal in earlier
+version of TLS by default. `s2n_config_set_alert_behavior()` can be called to
+force s2n-tls to treat pre-TLS1.3 warning alerts as not fatal, but that behavior
+is not recommended unless required for compatibility. In the past, attacks against
+TLS have involved manipulating the alert level to disguise fatal alerts as warnings.
+
+If s2n-tls encounters a fatal error, the next call to a write IO method will send
+a close_notify alert to the peer. Except for a few exceptions, s2n-tls does not
+send specific alerts in order to avoid leaking information that could be used for
+a sidechannel attack. To ensure that the alert is sent, `s2n_shutdown()` should
+be called after an error.
+
+### Performing the TLS Handshake
+
+Before application data can be sent or received, an application must perform a handshake
+to establish a TLS connection with the peer.
+
+To perform the handshake, call `s2n_negotiate()` until it either returns **S2N_SUCCESS**
+or returns **S2N_FAILURE** without a **S2N_ERR_T_BLOCKED** error.
+
+### Application Data
+
+After the TLS handshake, an application can send and receive encrypted data.
+
+Although most s2n-tls APIs are not thread-safe, `s2n_send()` and `s2n_recv()`
+may be called simultaneously from two different threads. This means that an
```

Comment:
Are you sure this is the case? Do we have a test that makes sure this behavior exists?

Reply:
I'm not sure if a test exists (those kinds of tests are very hard to write in a useful way), but we have advertised this as the expected behavior forever so if this is not the case, we have a serious bug. We enforce this by keeping the states completely separate: it's something we think about (or should be thinking about) every time we touch s2n_send or s2n_recv.